### ISSUE TITLE
Unstable NAR hash for `git` inputs using export-subst

### ISSUE THREAD

**Issue by mcwitt** [NixOS/nix#7596]
**Describe the bug**

When Flake inputs are git repos that use the `export-subst` feature in `.gitattributes` (for example, projects that use [python-versioneer](https://github.com/python-versioneer/python-versioneer)), the NAR hashes of these inputs can be unstable over time, leading to `NAR hash mismatch in input` errors.

**Steps To Reproduce**

Consider the flake

``` nix
{
  inputs.export-subst-minimal.url = "git+https://github.com/mcwitt/export-subst-minimal";
  inputs.export-subst-minimal.flake = false;
  outputs = _: { };
}
```

(the input [mcwitt/export-subst-minimal](https://github.com/mcwitt/export-subst-minimal) is a minimal example of a repo using `export-subst`)

The contents of the resulting store path are not reproducible:

```
$ nix flake lock
$ nix repl
…
nix-repl> :lf .
Added 8 variables.

nix-repl> inputs.export-subst-minimal.outPath
"/nix/store/iy7my74arsc4rgk9xrcs9mxrb0460x8d-source"

$ cat /nix/store/iy7my74arsc4rgk9xrcs9mxrb0460x8d-source/nondeterministic
 (HEAD -> main)
```

The contents will change for example when the pinned commit is no longer HEAD in main, leading to hash mismatch next time we fetch the input (on a different machine, or if the local cache is invalidated).

**Expected behavior**

Inputs using the `git+https` scheme should have stable NAR hashes. In particular, `git-archive` / `export-subst` actions should not be run, since these can lead to non-reproducibility as in the example above.

**`nix-env --version` output**

`nix-env (Nix) 2.11.1`

**Additional context**

- Original question: https://discourse.nixos.org/t/how-to-deal-with-unstable-flake-input-hashes-due-to-export-subst/24623
- Example workaround in nixpkgs (does not appear to work for Flake inputs): https://github.com/NixOS/nixpkgs/blob/971e45b76ab31f43891ca2ad5fb48fbd8578f654/pkgs/development/python-modules/limits/default.nix#L29-L34

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by roberth** [comment]
A different problem with unreproducible git fetchTree was discussed in [today's Nix Team meeting](https://discourse.nixos.org/t/2023-01-13-nix-team-meeting-minutes-23/24644/1). 

Another, somewhat-related data point for the `github:` type is that tree archives appear to receive less or no special processing.
Example: https://github.com/mcwitt/export-subst-minimal/archive/4e289bdc5e251b8aea07ba6cbc6b892450d95600.tar.gz

The tree hash could be stored in the lock file and it would also make subdir fetching more efficient.

**Comment by roberth** [comment]
Detailed notes: https://github.com/NixOS/nix/pull/4635#issuecomment-1381783001

**Comment by mcwitt** [comment]
> Another, somewhat-related data point for the github: type is that tree archives appear to receive less or no special processing.

I think `github:` type inputs are still susceptible to the same issue, though; in this case the substitution is done by github when generating the tree archive. The archives are apparently generated on demand and cached for a short period, so it's still possible to hit the `NAR hash mismatch in input` issue after some time. (Actually, I initially found this using `github:` and attempted a workaround by switching to `git+https:`, hoping the latter would be reproducible)

**Comment by Mic92** [comment]
If we were using `libgit2` would this give us more control over this?

**Comment by roberth** [comment]
It does, and we've been using libgit2 since Nix 2.20
- https://github.com/NixOS/nix/pull/9240

```
nix-repl> builtins.readFile (builtins.fetchTree { type = "git"; url = "https://github.com/mcwitt/export-subst-minimal"; } + "/nondeterministic")
"$Format:%d$\n"
```

This is what flakes does internally as well.

Solved since #9240
